blockchain/utxobackend.go

```diff
@@ -106,3 +110,79 @@ func (l *LevelDbUtxoBackend) FetchEntry(outpoint wire.OutPoint) (*UtxoEntry, err
 
 	return entry, nil
 }
+
+// dbPutUtxoEntry uses an existing database transaction to update the utxo
+// entry for the given outpoint based on the provided utxo entry state.  In
+// particular, the entry is only written to the database if it is marked as
+// modified, and if the entry is marked as spent it is removed from the
+// database.
+func (f *LevelDbUtxoBackend) dbPutUtxoEntry(dbTx database.Tx,
+	outpoint wire.OutPoint, entry *UtxoEntry) error {
+
+	// No need to update the database if the entry was not modified.
+	if entry == nil || !entry.isModified() {
+		return nil
+	}
+
+	// Remove the utxo entry if it is spent.
+	utxoBucket := dbTx.Metadata().Bucket(utxoSetBucketName)
+	if entry.IsSpent() {
+		key := outpointKey(outpoint)
+		err := utxoBucket.Delete(*key)
+		recycleOutpointKey(key)
+		if err != nil {
+			return err
+		}
+
+		return nil
+	}
+
+	// Serialize and store the utxo entry.
+	serialized, err := serializeUtxoEntry(entry)
+	if err != nil {
+		return err
+	}
+	key := outpointKey(outpoint)
+	err = utxoBucket.Put(*key, serialized)
+	// NOTE: The key is intentionally not recycled here since the database
+	// interface contract prohibits modifications.  It will be garbage collected
+	// normally when the database is done with it.
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// PutUtxos atomically updates the UTXO set with the entries from the provided
+// map along with the current state.
+func (l *LevelDbUtxoBackend) PutUtxos(utxos map[wire.OutPoint]*UtxoEntry,
+	state *utxoSetState) error {
+
+	// Update the database with the provided entries and UTXO set state.
+	//
+	// It is important that the UTXO set state is always updated in the same
+	// database transaction as the utxo set itself so that it is always in sync.
+	err := l.db.Update(func(dbTx database.Tx) error {
+		for outpoint, entry := range utxos {
+			// Write the entry to the database.
+			err := l.dbPutUtxoEntry(dbTx, outpoint, entry)
+			if err != nil {
+				return err
+			}
+		}
+
+		// Update the UTXO set state in the database.
+		return dbTx.Metadata().Put(utxoSetStateKeyName,
+			serializeUtxoSetState(state))
+	})
+	if err != nil {
+		return err
+	}
+
+	// Flush the UTXO database to disk.  This is necessary in the case that the
```

Comment:
This actually is accurate, it is only _required_ when the state was just initialized.  In any other case, the block database is always flushed before the cache is flushed, so the cache will always be ahead of the block database and, therefore, can recover on startup.  This also won't be needed once we move to the new backend since this database caching layer won't exist anymore.

Reply:
Make sense. Thanks for the clarification.